### ISSUE TITLE
eslint run with --stdin does not correctly set the exitCode

### ISSUE THREAD

**Issue by shumphrey** [eslint/eslint#6677]
**What version of ESLint are you using?**
v3.0.1

**What parser (default, Babel-ESLint, etc.) are you using?**
default

**Please show your full configuration:**

**What did you do? Please include the actual source code causing the issue.**
cat file | eslint --stdin
echo $?

**What did you expect to happen?**
I expect 1 to be printed

**What actually happened? Please include the actual, raw output from ESLint.**
0 is printed

This is due to exitCode being set in a callback, but process.exitCode being set outside the callback and thus executing before exitCode is set.
I have a fix for this locally and would like to submit a PR.

**Comment by platinumazure** [comment]
Just to be clear, is this something you expect because you know your file has lint errors in it, but the wrong exit code is being sent along? (If your file had no errors, then I would definitely expect an exit code of zero)

**Comment by shumphrey** [comment]
Yes, the file has errors.

    eslint file && echo $?

produces 1

    cat file | eslint --stdin
    echo $?

produces 0

**Comment by platinumazure** [comment]
Awesome. Appreciate the clarification- without the file and configuration, I had to be sure.

It does sound like a bug to me. Hopefully we can get this issue accepted quickly.

**Comment by shumphrey** [comment]
18fbc4b86ac9fb84fb370c68b9a60d9b70cafa66 introduced this bug.

Previously it did
```javascript
process.on("exit", function() {
    process.exit(exitCode);
});
```
Where as now it does
```javascript
process.exitCode = exitCode
```

The former accounted for the async nature of <code>--stdin</code> where as the latter does not.

**Comment by platinumazure** [comment]
Pretty sure the commit was only done because we thought `process.exitCode = exitCode` was just cleaner/more efficient once we dropped older Node versions. If it needs to be reverted in order to get this working again, I have no problem with that.

**Comment by shumphrey** [comment]
I don't think it needs reverting.
It just needs the line <code>process.exitCode = exitCode</code> adding to the end of stdin callback

**Comment by platinumazure** [comment]
@shumphrey Thanks for the guidance.

In the same file, we have special handling for the `--init` flag. Assuming that callback is also asynchronous, should we also consider adding `process.exitCode = exitCode;` to the end of that callback? Thanks!

**Comment by shumphrey** [comment]
I'll check

**Comment by mysticatea** [comment]
I think we can remove the local variable `exitCode`, then use `process.exitCode` directly.

**Comment by shumphrey** [comment]
I agree

**Comment by platinumazure** [comment]
@mysticatea Can we accept this issue?

**Comment by shumphrey** [comment]
For what it's worth, the init stuff seems to be synchronous and I couldn't quickly get it to run the callback.
Anyway, I think @mysticatea is correct. Just get rid of the exitCode variable and have the blocks set process.exitCode as needed

**Comment by mysticatea** [comment]
I confirmed.

```
$ echo var a | eslint --stdin --rule no-var:error --no-eslintrc --no-ignore && echo bad

<text>
  1:1  error  Unexpected var, use let or const instead  no-var

✖ 1 problem (1 error, 0 warnings)

bad

$ eslint test.js --rule no-var:error --no-eslintrc --no-ignore && echo bad

/path/to/test.js
  1:1  error  Unexpected var, use let or const instead  no-var

✖ 1 problem (1 error, 0 warnings)
```
